File: Data Structures & Algorithms/combinations-of-a-phone-number/submission-2.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []
        digitToChar = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jkl",
            "6": "mno",
            "7": "qprs",
            "8": "tuv",
            "9": "wxyz",
        }        
        res = []
        def dfs(i: int, cur_res: str) -> None:
            if i == len(digits):
                res.append(cur_res)
                return
            
            for ch in digitToChar[digits[i]]:
                cur_res += ch
                dfs(i + 1, cur_res)
                cur_res = cur_res[:-1]
            
        dfs(0, "")
        return res

File: Data Structures & Algorithms/generate-parentheses/submission-2.cpp
class Solution {
public:
    vector<string> res;
    void dfs(int openingCount, int closingCount, int n, string& curRes) {
        if (openingCount > n || closingCount > openingCount) {
            return;
        }
        if (openingCount + closingCount == 2*n){
            res.push_back(curRes);
            return;
        }
        curRes.push_back('(');
        dfs(openingCount + 1, closingCount, n, curRes);
        curRes.pop_back();
        curRes.push_back(')');
        dfs(openingCount, closingCount + 1, n, curRes);
        curRes.pop_back();
    }
    vector<string> generateParenthesis(int n) {
        string curRes = "";
        dfs(0, 0, n, curRes);
        return res;
    }
};
